sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressionsJoni.scala

```diff
@@ -0,0 +1,471 @@
+            |}
+      """.stripMargin)
+  }
+}
+
+case class LikeAnyJoni(child: Expression, patterns: Seq[UTF8String]) extends LikeAnyJoniBase {
+  override def isNotSpecified: Boolean = false
+  override protected def withNewChildInternal(newChild: Expression): LikeAnyJoni =
+    copy(child = newChild)
+}
+
+case class NotLikeAnyJoni(child: Expression, patterns: Seq[UTF8String]) extends LikeAnyJoniBase {
+  override def isNotSpecified: Boolean = true
+  override protected def withNewChildInternal(newChild: Expression): NotLikeAnyJoni =
+    copy(child = newChild)
+}
+
+
+// scalastyle:off line.contains.tab
+@ExpressionDescription(
+  usage = "str _FUNC_ regexp - Returns true if `str` matches `regexp`, or false otherwise.",
+  arguments = """
+    Arguments:
+      * str - a string expression
+      * regexp - a string expression. The regex string should be a Java regular expression.
+
+          Since Spark 2.0, string literals (including regex patterns) are unescaped in our SQL
+          parser. For example, to match "\abc", a regular expression for `regexp` can be
+          "^\\abc$".
+
+          There is a SQL config 'spark.sql.parser.escapedStringLiterals' that can be used to
+          fallback to the Spark 1.6 behavior regarding string literal parsing. For example,
+          if the config is enabled, the `regexp` that can match "\abc" is "^\abc$".
+  """,
+  examples = """
+    Examples:
+      > SET spark.sql.parser.escapedStringLiterals=true;
+      spark.sql.parser.escapedStringLiterals	true
+      > SELECT '%SystemDrive%\Users\John' _FUNC_ '%SystemDrive%\\Users.*';
+      true
+      > SET spark.sql.parser.escapedStringLiterals=false;
+      spark.sql.parser.escapedStringLiterals	false
+      > SELECT '%SystemDrive%\\Users\\John' _FUNC_ '%SystemDrive%\\\\Users.*';
+      true
+  """,
+  note = """
+    Use LIKE to match with simple string pattern.
+  """,
+  since = "1.0.0")
+// scalastyle:on line.contains.tab
+case class RLikeJoni(left: Expression, right: Expression) extends StringRegexExpressionJoni {
+
+  override def escape(v: Array[Byte]): Array[Byte] = v
+  override def matches(regex: Regex, input: Array[Byte]): Boolean = {
+    regex.matcher(input).search(0, input.length, Option.DEFAULT) > -1
+  }
+  override def toString: String = s"$left RLIKE_JONI $right"
+
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val regexClass = classOf[Regex].getName
+    val optionClass = classOf[Option].getName
+    val encodingClass = classOf[UTF8Encoding].getName
+    val syntaxClass = classOf[Syntax].getName
+    val regex = ctx.freshName("regex")
+
+    if (right.foldable) {
+      val rVal = right.eval()
+      if (rVal != null) {
+        val tmp =
+          StringEscapeUtils.escapeJava(rVal.asInstanceOf[UTF8String].toString())
+        val pattern = ctx.addMutableState(regexClass, regex,
+          v => s"""
+                    byte[] pattern = UTF8String.fromString("${tmp}").getBytes();
```

Comment:
codegen will fail for redefinition of local variable "pattern".

```sql
select 
  case when columnA rlike '.*abcd.*' then 1 
       when columnA rlike '.*abcdc.*' then 2  else 0 end as a
from tableA
```

Reply:
Could you point out exactly where the problem is caused? thanks~

Reply:
fix this issue